_source/_posts/2022-06-01-cloud-native-java-microservices-with-istio.md

```diff
@@ -0,0 +1,604 @@
+        maxRequestsPerConnection: 10
+        maxRetries: 5
+    outlierDetection:
+      consecutive5xxErrors: 5
+      interval: 30s
+      baseEjectionTime: 60s
+  subsets:
+    - name: v1
+      labels:
+        version: "v1"
+```
+
+It also has the below virtual service that defines the route. You could also use virtual services to do traffic split between two versions of the same app, among other things.
+
+```yaml
+apiVersion: networking.istio.io/v1beta1
+kind: VirtualService
+metadata:
+  name: invoice-virtualservice
+  namespace: jhipster
+spec:
+  hosts:
+    - invoice
+  http:
+    - route:
+        - destination:
+            host: invoice
+            subset: "v1"
+          weight: 100
+      retries:
+        attempts: 3
+        perTryTimeout: 2s
+```
+
+The gateway is defined for the store application as it is our GUI as well.
+
+```yaml
+apiVersion: networking.istio.io/v1beta1
+kind: Gateway
+metadata:
+  name: store-gateway
+  namespace: jhipster
+  labels:
+    gateway: store-gateway
+    istio: ingressgateway
+spec:
+  selector:
+    istio: ingressgateway
+  servers:
+    - port:
+        number: 80
+        name: http
+        protocol: HTTP
+      hosts:
+        - store.jhipster.34.76.233.160.nip.io
+    - port:
+        number: 80
+        name: http2
+        protocol: HTTP2
+      hosts:
+        - store.jhipster.34.76.233.160.nip.io
+---
+apiVersion: networking.istio.io/v1beta1
+kind: VirtualService
+metadata:
+  name: store-gw-virtualservice
+  namespace: jhipster
+  labels:
+    service: store-gw-virtualservice
+spec:
+  hosts:
+    - store.jhipster.34.76.233.160.nip.io
+  gateways:
+    - store-gateway
+  http:
+    - match:
+        - uri:
+            prefix: /services/invoice/
+      rewrite:
+        uri: /
+      route:
+        - destination:
+            host: invoice
+    - match:
+        - uri:
+            prefix: /services/notification/
+      rewrite:
+        uri: /
+      route:
+        - destination:
+            host: notification
+    - match:
+        - uri:
+            prefix: /services/product/
+      rewrite:
+        uri: /
+      route:
+        - destination:
+            host: product
+    - route:
+        - destination:
+            host: store
+```
+
+As you can see, there are also many useful commands printed on the console that you can use to do the deployment.
+
+### Deploy to GKE
+
+We are ready to deploy now. First, we need to build and push the images to the registry. We can use the handy [Jib](https://github.com/GoogleContainerTools/jib) commands provided by JHipster. Navigate to each of the microservice folders and run the command below.
+
+```bash
+cd store && ./gradlew bootJar -Pprod jib -Djib.to.image=deepu105/store
+cd invoice && ./gradlew bootJar -Pprod jib -Djib.to.image=deepu105/invoice
+cd notification && ./gradlew bootJar -Pprod jib -Djib.to.image=deepu105/notification
+cd product && ./gradlew bootJar -Pprod jib -Djib.to.image=deepu105/product
```

Comment:
These commands have your Docker Hub username in them. Should the reader change them to use their own?